File: test/e2e/examples.test.ts
import { afterEach, beforeEach, describe, expect, test } from "bun:test";
import { realpathSync } from "node:fs";
import { readFile } from "node:fs/promises";
import { join } from "node:path";
import { ExampleHarness, type RpcCall, type RunResult } from "./examples-harness";

let harness: ExampleHarness;

beforeEach(async () => {
  harness = await ExampleHarness.create();
});

afterEach(async () => {
  await harness.close();
});

function successful(result: RunResult): RpcCall[] {
  expect(result.exitCode, result.stderr).toBe(0);
  expect(result.stderr).toBe("");
  return result.calls;
}

function notifications(calls: RpcCall[]): Record<string, unknown>[] {
  return calls.filter((call) => call.method === "notification.show").map((call) => call.params);
}

function titles(calls: RpcCall[]): unknown[] {
  return notifications(calls).map((params) => params.title);
}

describe("shipped examples through isolated CLI and fake Herdr", () => {
  test("branch-check covers clean, changed, existing, and available outcomes", async () => {
    const scenarios: {
      inputs: Record<string, string>;
      env: Record<string, string>;
      title: string;
    }[] = [
      {
        inputs: { mode: "status" },
        env: { HWF_E2E_GIT_DIRTY: "0" },
        title: "working tree clean",
      },
      {
        inputs: { mode: "status" },
        env: { HWF_E2E_GIT_DIRTY: "1" },
        title: "working tree changed",
      },
      {
        inputs: { mode: "branch", branch: "main" },
        env: { HWF_E2E_BRANCH_EXISTS: "1" },
        title: "branch exists",
      },
      {
        inputs: { mode: "branch", branch: "new-feature" },
        env: { HWF_E2E_BRANCH_EXISTS: "0" },
        title: "branch available",
      },
    ];

    for (const scenario of scenarios) {
      const calls = successful(await harness.run("branch-check", scenario.inputs, scenario.env));
      expect(notifications(calls).map((params) => params.title)).toEqual([scenario.title]);
    }
  });

  test("prompt-enhance uses the configured custom agent and this platform's clipboard", async () => {
    const calls = successful(
      await harness.run(
        "prompt-enhance",
        { target: "deterministic", text: "fix it" },
        { WAYLAND_DISPLAY: "wayland-e2e" },
      ),
    );

    const expected = process.platform === "darwin" ? "pbcopy" : "wl-copy";
    expect(await readFile(harness.clipboard, "utf8")).toBe(`${expected}:refined prompt`);
    expect(calls.find((call) => call.method === "agent.start")?.params).toMatchObject({
      kind: "custom",
    });
    expect(titles(calls)).toEqual(["enhancing prompt", "prompt ready"]);
    expect(calls.some((call) => call.method === "pane.close")).toBe(true);
  });

  // macOS never reaches this branch: the workflow guards it on context.platform.
  test.skipIf(process.platform === "darwin")(
    "prompt-enhance falls back to xclip when no Wayland display is set",
    async () => {
      const calls = successful(
        await harness.run(
          "prompt-enhance",
          { target: "deterministic", text: "fix it" },
          { WAYLAND_DISPLAY: "" },
        ),
      );

      expect(await readFile(harness.clipboard, "utf8")).toBe("xclip:refined prompt");
      expect(titles(calls)).toEqual(["enhancing prompt", "prompt ready"]);
    },
  );

  test("handoff preserves the target and cleans up at the selected granularity", async () => {
    for (const placement of ["tab", "beside", "below"]) {
      for (const closeSource of ["keep", "close"]) {
        const calls = successful(
          await harness.run("handoff", {
            target: "deterministic",
            focus: "",
            placement,
            close_source: closeSource,
          }),
        );
        const starts = calls.filter((call) => call.method === "agent.start");
        expect(starts).toHaveLength(2);
        expect(await readFile(join(harness.repoRoot, ".hwf", "tmp", "handoff.md"), "utf8")).toBe(
          "deterministic handoff",
        );
        const targetPane = String(starts[1]?.params.pane_id);
        const sourcePaneCloses = calls.filter(
          (call) => call.method === "pane.close" && call.params.pane_id === "w1:p1",
        );
        const sourceTabCloses = calls.filter(
          (call) => call.method === "tab.close" && call.params.tab_id === "w1:t1",
        );

        expect(
          calls.some((call) => call.method === "pane.close" && call.params.pane_id === targetPane),
        ).toBe(false);
        if (closeSource === "keep") {
          expect(sourcePaneCloses).toHaveLength(0);
          expect(sourceTabCloses).toHaveLength(0);
        } else if (placement === "tab") {
          expect(sourceTabCloses).toHaveLength(1);
          expect(sourcePaneCloses).toHaveLength(0);
        } else {
          expect(sourcePaneCloses).toHaveLength(1);
          expect(sourceTabCloses).toHaveLength(0);
        }

        if (placement === "tab") {
          expect(calls.filter((call) => call.method === "tab.create")).toHaveLength(2);
        } else {
          const expectedDirection = placement === "beside" ? "right" : "down";
          expect(
            calls.some(
              (call) =>
                call.method === "pane.split" &&
                call.params.direction === expectedDirection &&
                call.params.target_pane_id === "w1:p1",
            ),
          ).toBe(true);
        }
      }
    }
  });

  test("worktree create runs the full layout with a per-pane agent name", async () => {
    const calls = successful(
      await harness.run("worktree", { mode: "create", ref: "main", branch: "feat-x" }),
    );
    const created = calls.find((call) => call.method === "worktree.create");
    expect(created?.params).toMatchObject({
      cwd: realpathSync(harness.repoRoot),
      branch: "feat-x",
      base: "main",
      label: "feat-x",
      focus: true,
    });
    const rename = calls.find((call) => call.method === "tab.rename");
    expect(rename?.params.label).toBe("work");
    const started = calls.find((call) => call.method === "agent.start");
    expect(started?.params.kind).toBe("claude");
    expect(started?.params.name).toBe(
      `claude-${String(started?.params.pane_id)}`.replace(/[^A-Za-z0-9-]/g, "-"),
    );
    expect(calls.some((call) => call.method === "tab.focus")).toBe(true);
    expect(notifications(calls).map((params) => params.title)).toEqual(["Worktree ready"]);
  });

  test("worktree open targets an existing branch through the same layout", async () => {
    const calls = successful(
      await harness.run("worktree", { mode: "open", worktree: "feature-seed" }),
    );
    const opened = calls.find((call) => call.method === "worktree.open");
    expect(opened?.params).toMatchObject({
      cwd: realpathSync(harness.repoRoot),
      branch: "feature-seed",
      label: "feature-seed",
      focus: true,
    });
    expect(calls.some((call) => call.method === "worktree.create")).toBe(false);
    expect(notifications(calls).map((params) => params.title)).toEqual(["Worktree ready"]);
  });

  test("review-gate notifies on APPROVE and fails the run on REJECT", async () => {
    const approved = successful(
      await harness.run(
        "review-gate",
        { reviewer: "deterministic" },
        { HWF_E2E_GIT_DIRTY: "1", HWF_E2E_REVIEW_VERDICT: "APPROVE" },
      ),
    );
    expect(titles(approved)).toEqual(["review approved"]);
    expect(approved.some((call) => call.method === "pane.close")).toBe(true);

    const rejected = await harness.run(
      "review-gate",
      { reviewer: "deterministic" },
      { HWF_E2E_GIT_DIRTY: "1", HWF_E2E_REVIEW_VERDICT: "REJECT" },
    );
    expect(rejected.exitCode).not.toBe(0);
    expect(rejected.stderr).toContain("one finding, reported above");
    expect(titles(rejected.calls)).toEqual([]);
  });

  test("review-gate skips the reviewer when the diff is empty", async () => {
    const calls = successful(
      await harness.run("review-gate", { reviewer: "deterministic" }, { HWF_E2E_GIT_DIRTY: "0" }),
    );
    expect(titles(calls)).toEqual(["nothing to review"]);
    expect(calls.some((call) => call.method === "agent.start")).toBe(false);
  });

  test("adversarial-revise runs the revision step only on the REVISE verdict", async () => {
    const scenarios = [
      { verdict: "APPROVE", starts: 2, title: "proposal approved" },
      { verdict: "REVISE", starts: 3, title: "proposal revised" },
    ];
    for (const scenario of scenarios) {
      const calls = successful(
        await harness.run(
          "adversarial-revise",
          { task: "ship a thing", author: "deterministic", critic: "deterministic" },
          { HWF_E2E_CRITIQUE_VERDICT: scenario.verdict },
        ),
      );
      expect(calls.filter((call) => call.method === "agent.start")).toHaveLength(scenario.starts);
      expect(titles(calls)).toEqual([scenario.title]);
    }
  });

  test("remote-branch-log resolves the branch choice from the chosen remote", async () => {
    for (const remote of ["origin", "upstream"]) {
      const calls = successful(
        await harness.run("remote-branch-log", { remote, branch: `${remote}/release` }),
      );
      expect(notifications(calls)).toEqual([
        {
          title: `recent commits on ${remote}/release`,
          body: `abc1234 seed commit on ${remote}/release\n`,
          sound: "done",
        },
      ]);
    }

    const crossed = await harness.run("remote-branch-log", {
      remote: "origin",
      branch: "upstream/release",
    });
    expect(crossed.exitCode).not.toBe(0);
    expect(crossed.stderr).toContain("must be one of: origin/main, origin/release");
  });

  test("worktree open skips agent.start when the reopened pane already has one", async () => {
    const calls = successful(
      await harness.run(
        "worktree",
        { mode: "open", worktree: "feature-seed" },
        { HWF_E2E_AGENT_ON_OPEN: "1" },
      ),
    );

    expect(calls.some((call) => call.method === "worktree.open")).toBe(true);
    expect(calls.filter((call) => call.method === "agent.start")).toHaveLength(0);
    expect(calls.some((call) => call.method === "tab.focus")).toBe(true);
    expect(titles(calls)).toEqual(["Worktree ready"]);
  });

  test("worktree delete removes the checkout and reports the branch outcome", async () => {
    const scenarios: { scope: string; env: Record<string, string>; body: string }[] = [
      {
        scope: "worktree-only",
        env: {},
        body: "removed the feature-seed worktree; branch kept",
      },
      {
        scope: "worktree-and-branch",
        env: {},
        body: "removed the feature-seed worktree and its branch",
      },
      {
        scope: "worktree-and-branch",
        env: { HWF_E2E_BRANCH_UNMERGED: "1" },
        body: "removed the feature-seed worktree; branch kept, it is not merged (git branch -D feature-seed)",
      },
    ];
    for (const scenario of scenarios) {
      const calls = successful(
        await harness.run(
          "worktree",
          { mode: "delete", worktree: "feature-seed", scope: scenario.scope },
          scenario.env,
        ),
      );
      expect(calls.some((call) => call.method === "worktree.create")).toBe(false);
      const shown = notifications(calls);
      expect(shown.map((params) => params.title)).toEqual(["Worktree deleted"]);
      expect(shown[0]?.body).toBe(scenario.body);
    }
  });
});

File: e2e/examples_test.go
package e2e_test

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
)

func TestBranchCheckCoversCleanChangedExistingAndAvailableOutcomes(t *testing.T) {
	h := newExampleHarness(t)
	scenarios := []struct {
		inputs map[string]string
		env    map[string]string
		title  string
	}{
		{
			inputs: map[string]string{"mode": "status"},
			env:    map[string]string{"HWF_E2E_GIT_DIRTY": "0"},
			title:  "working tree clean",
		},
		{
			inputs: map[string]string{"mode": "status"},
			env:    map[string]string{"HWF_E2E_GIT_DIRTY": "1"},
			title:  "working tree changed",
		},
		{
			inputs: map[string]string{"mode": "branch", "branch": "main"},
			env:    map[string]string{"HWF_E2E_BRANCH_EXISTS": "1"},
			title:  "branch exists",
		},
		{
			inputs: map[string]string{"mode": "branch", "branch": "new-feature"},
			env:    map[string]string{"HWF_E2E_BRANCH_EXISTS": "0"},
			title:  "branch available",
		},
	}

	for _, scenario := range scenarios {
		calls := runOK(t, h, "branch-check", scenario.inputs, scenario.env)
		got := make([]any, 0, len(notifications(calls)))
		for _, params := range notifications(calls) {
			got = append(got, params["title"])
		}
		if len(got) != 1 || got[0] != scenario.title {
			t.Fatalf("scenario %+v titles = %v want [%q]", scenario.inputs, got, scenario.title)
		}
	}
}

func TestPromptEnhanceUsesConfiguredCustomAgentAndPlatformClipboard(t *testing.T) {
	h := newExampleHarness(t)
	calls := runOK(t, h, "prompt-enhance", map[string]string{"target": "deterministic", "text": "fix it"}, map[string]string{"WAYLAND_DISPLAY": "wayland-e2e"})

	expected := "wl-copy"
	if runtime.GOOS == "darwin" {
		expected = "pbcopy"
	}
	clip, err := os.ReadFile(h.clipboard)
	if err != nil {
		t.Fatal(err)
	}
	if string(clip) != expected+":refined prompt" {
		t.Fatalf("clipboard = %q want %q", clip, expected+":refined prompt")
	}
	var startParams map[string]any
	for _, call := range calls {
		if call.Method == "agent.start" {
			startParams = call.Params
			break
		}
	}
	if startParams == nil || startParams["kind"] != "custom" {
		t.Fatalf("agent.start params = %#v", startParams)
	}
	if got := titles(calls); len(got) != 1 || got[0] != "enhancing prompt" {
		t.Fatalf("titles = %v", got)
	}
	foundClose := false
	for _, call := range calls {
		if call.Method == "pane.close" {
			foundClose = true
			break
		}
	}
	if !foundClose {
		t.Fatal("expected pane.close")
	}
}

func TestPromptEnhanceFallsBackToXclipWhenNoWaylandDisplayIsSet(t *testing.T) {
	if runtime.GOOS == "darwin" {
		t.Skip("macOS never reaches this branch: the workflow guards it on context.platform")
	}
	h := newExampleHarness(t)
	calls := runOK(t, h, "prompt-enhance", map[string]string{"target": "deterministic", "text": "fix it"}, map[string]string{"WAYLAND_DISPLAY": ""})
	clip, err := os.ReadFile(h.clipboard)
	if err != nil {
		t.Fatal(err)
	}
	if string(clip) != "xclip:refined prompt" {
		t.Fatalf("clipboard = %q", clip)
	}
	if got := titles(calls); len(got) != 1 || got[0] != "enhancing prompt" {
		t.Fatalf("titles = %v", got)
	}
}

func TestHandoffPreservesTargetAndCleansUpAtSelectedGranularity(t *testing.T) {
	h := newExampleHarness(t)
	for _, placement := range []string{"tab", "beside", "below"} {
		for _, closeSource := range []string{"keep", "close"} {
			calls := runOK(t, h, "handoff", map[string]string{
				"target":       "deterministic",
				"focus":        "",
				"placement":    placement,
				"close_source": closeSource,
			}, nil)

			starts := filterCalls(calls, "agent.start")
			if len(starts) != 2 {
				t.Fatalf("placement=%s close=%s agent.start count = %d", placement, closeSource, len(starts))
			}
			handoff, err := os.ReadFile(filepath.Join(h.repoRoot, ".hwf", "tmp", "handoff.md"))
			if err != nil {
				t.Fatal(err)
			}
			if string(handoff) != "deterministic handoff" {
				t.Fatalf("handoff = %q", handoff)
			}
			targetPane := fmtString(starts[1].Params["pane_id"])
			sourcePaneCloses := filterCallsWhere(calls, "pane.close", func(p map[string]any) bool {
				return fmtString(p["pane_id"]) == "w1:p1"
			})
			sourceTabCloses := filterCallsWhere(calls, "tab.close", func(p map[string]any) bool {
				return fmtString(p["tab_id"]) == "w1:t1"
			})
			targetClosed := false
			for _, call := range calls {
				if call.Method == "pane.close" && fmtString(call.Params["pane_id"]) == targetPane {
					targetClosed = true
				}
			}
			if targetClosed {
				t.Fatalf("target pane %s was closed", targetPane)
			}
			assertHandoffCloseBehavior(t, placement, closeSource, sourcePaneCloses, sourceTabCloses)
			assertHandoffPlacementCalls(t, calls, placement)
		}
	}
}

func assertHandoffCloseBehavior(t *testing.T, placement, closeSource string, sourcePaneCloses, sourceTabCloses []rpcCall) {
	t.Helper()
	if closeSource == "keep" {
		if len(sourcePaneCloses) != 0 || len(sourceTabCloses) != 0 {
			t.Fatalf("keep: pane closes=%d tab closes=%d", len(sourcePaneCloses), len(sourceTabCloses))
		}
		return
	}
	if placement == "tab" {
		if len(sourceTabCloses) != 1 || len(sourcePaneCloses) != 0 {
			t.Fatalf("tab close: pane closes=%d tab closes=%d", len(sourcePaneCloses), len(sourceTabCloses))
		}
		return
	}
	if len(sourcePaneCloses) != 1 || len(sourceTabCloses) != 0 {
		t.Fatalf("%s close: pane closes=%d tab closes=%d", placement, len(sourcePaneCloses), len(sourceTabCloses))
	}
}

func assertHandoffPlacementCalls(t *testing.T, calls []rpcCall, placement string) {
	t.Helper()
	if placement == "tab" {
		if len(filterCalls(calls, "tab.create")) != 2 {
			t.Fatal("expected two tab.create calls")
		}
		return
	}
	expectedDirection := "right"
	if placement == "below" {
		expectedDirection = "down"
	}
	for _, call := range calls {
		if call.Method == "pane.split" &&
			fmtString(call.Params["direction"]) == expectedDirection &&
			fmtString(call.Params["target_pane_id"]) == "w1:p1" {
			return
		}
	}
	t.Fatalf("missing pane.split direction=%s", expectedDirection)
}

func TestWorktreeCreateRunsFullLayoutWithPerPaneAgentName(t *testing.T) {
	h := newExampleHarness(t)
	calls := runOK(t, h, "worktree", map[string]string{
		"mode": "create", "ref": "main", "branch": "feat-x",
	}, nil)
	created := findCall(calls, "worktree.create")
	if created == nil {
		t.Fatal("missing worktree.create")
	}
	wantCwd := repoRealpath(t, h.repoRoot)
	if fmtString(created.Params["cwd"]) != wantCwd ||
		fmtString(created.Params["branch"]) != "feat-x" ||
		fmtString(created.Params["base"]) != "main" ||
		fmtString(created.Params["label"]) != "feat-x" ||
		created.Params["focus"] != true {
		t.Fatalf("worktree.create params = %#v", created.Params)
	}
	rename := findCall(calls, "tab.rename")
	if rename == nil || fmtString(rename.Params["label"]) != "work" {
		t.Fatalf("tab.rename = %#v", rename)
	}
	started := findCall(calls, "agent.start")
	if started == nil || fmtString(started.Params["kind"]) != "claude" {
		t.Fatalf("agent.start = %#v", started)
	}
	paneID := fmtString(started.Params["pane_id"])
	wantName := sanitizeAgentName("claude-" + paneID)
	if fmtString(started.Params["name"]) != wantName {
		t.Fatalf("agent name = %q want %q", started.Params["name"], wantName)
	}
	if findCall(calls, "tab.focus") == nil {
		t.Fatal("missing tab.focus")
	}
	if got := titles(calls); len(got) != 0 {
		t.Fatalf("titles = %v", got)
	}
}

func TestWorktreeOpenTargetsExistingBranchThroughSameLayout(t *testing.T) {
	h := newExampleHarness(t)
	calls := runOK(t, h, "worktree", map[string]string{
		"mode": "open", "worktree": "feature-seed",
	}, nil)
	opened := findCall(calls, "worktree.open")
	if opened == nil {
		t.Fatal("missing worktree.open")
	}
	wantCwd := repoRealpath(t, h.repoRoot)
	if fmtString(opened.Params["cwd"]) != wantCwd ||
		fmtString(opened.Params["branch"]) != "feature-seed" ||
		fmtString(opened.Params["label"]) != "feature-seed" ||
		opened.Params["focus"] != true {
		t.Fatalf("worktree.open params = %#v", opened.Params)
	}
	if findCall(calls, "worktree.create") != nil {
		t.Fatal("unexpected worktree.create")
	}
	if got := titles(calls); len(got) != 0 {
		t.Fatalf("titles = %v", got)
	}
}

func TestReviewGateNotifiesOnApproveAndFailsRunOnReject(t *testing.T) {
	h := newExampleHarness(t)
	approved := runOK(t, h, "review-gate", map[string]string{"reviewer": "deterministic"}, map[string]string{"HWF_E2E_GIT_DIRTY": "1", "HWF_E2E_REVIEW_VERDICT": "APPROVE"})
	if got := titles(approved); len(got) != 1 || got[0] != "review approved" {
		t.Fatalf("approved titles = %v", got)
	}
	if findCall(approved, "pane.close") == nil {
		t.Fatal("expected pane.close on approve")
	}

	rejected, err := h.run(
		"review-gate",
		map[string]string{"reviewer": "deterministic"},
		map[string]string{"HWF_E2E_GIT_DIRTY": "1", "HWF_E2E_REVIEW_VERDICT": "REJECT"},
	)
	if err != nil {
		t.Fatal(err)
	}
	if rejected.ExitCode == 0 {
		t.Fatal("expected non-zero exit on reject")
	}
	if !strings.Contains(rejected.Stderr, "one finding, reported above") {
		t.Fatalf("stderr = %q", rejected.Stderr)
	}
	if len(titles(rejected.Calls)) != 0 {
		t.Fatalf("reject titles = %v", titles(rejected.Calls))
	}
}

func TestReviewGateSkipsReviewerWhenDiffIsEmpty(t *testing.T) {
	h := newExampleHarness(t)
	calls := runOK(t, h, "review-gate", map[string]string{"reviewer": "deterministic"}, map[string]string{"HWF_E2E_GIT_DIRTY": "0"})
	if got := titles(calls); len(got) != 1 || got[0] != "nothing to review" {
		t.Fatalf("titles = %v", got)
	}
	if findCall(calls, "agent.start") != nil {
		t.Fatal("unexpected agent.start")
	}
}

func TestAdversarialReviseRunsRevisionStepOnlyOnReviseVerdict(t *testing.T) {
	h := newExampleHarness(t)
	scenarios := []struct {
		verdict string
		starts  int
		title   string
	}{
		{verdict: "APPROVE", starts: 2, title: "proposal approved"},
		{verdict: "REVISE", starts: 3, title: "proposal revised"},
	}
	for _, scenario := range scenarios {
		calls := runOK(t, h, "adversarial-revise", map[string]string{"task": "ship a thing", "author": "deterministic", "critic": "deterministic"}, map[string]string{"HWF_E2E_CRITIQUE_VERDICT": scenario.verdict})
		if len(filterCalls(calls, "agent.start")) != scenario.starts {
			t.Fatalf("verdict=%s starts = %d want %d", scenario.verdict, len(filterCalls(calls, "agent.start")), scenario.starts)
		}
		if got := titles(calls); len(got) != 1 || got[0] != scenario.title {
			t.Fatalf("verdict=%s titles = %v", scenario.verdict, got)
		}
	}
}

func TestRemoteBranchLogResolvesBranchChoiceFromChosenRemote(t *testing.T) {
	h := newExampleHarness(t)
	for _, remote := range []string{"origin", "upstream"} {
		calls := runOK(t, h, "remote-branch-log", map[string]string{
			"remote": remote,
			"branch": remote + "/release",
		}, nil)
		shown := notifications(calls)
		if len(shown) != 1 {
			t.Fatalf("remote=%s notifications = %d", remote, len(shown))
		}
		if fmtString(shown[0]["title"]) != "recent commits on "+remote+"/release" {
			t.Fatalf("title = %v", shown[0]["title"])
		}
		if body, ok := shown[0]["body"].(string); !ok || body != "abc1234 seed commit on "+remote+"/release\n" {
			t.Fatalf("body = %q", shown[0]["body"])
		}
		if fmtString(shown[0]["sound"]) != "done" {
			t.Fatalf("sound = %v", shown[0]["sound"])
		}
	}

	crossed, err := h.run("remote-branch-log", map[string]string{
		"remote": "origin",
		"branch": "upstream/release",
	}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if crossed.ExitCode == 0 {
		t.Fatal("expected failure for crossed remote/branch")
	}
	if !strings.Contains(crossed.Stderr, "must be one of: origin/main, origin/release") {
		t.Fatalf("stderr = %q", crossed.Stderr)
	}
}

func TestWorktreeOpenSkipsAgentStartWhenReopenedPaneAlreadyHasOne(t *testing.T) {
	h := newExampleHarness(t)
	calls := runOK(t, h, "worktree", map[string]string{"mode": "open", "worktree": "feature-seed"}, map[string]string{"HWF_E2E_AGENT_ON_OPEN": "1"})
	if findCall(calls, "worktree.open") == nil {
		t.Fatal("missing worktree.open")
	}
	if len(filterCalls(calls, "agent.start")) != 0 {
		t.Fatal("unexpected agent.start")
	}
	if findCall(calls, "tab.focus") == nil {
		t.Fatal("missing tab.focus")
	}
	if got := titles(calls); len(got) != 0 {
		t.Fatalf("titles = %v", got)
	}
}

func TestWorktreeDeleteRemovesCheckoutAndReportsBranchOutcome(t *testing.T) {
	h := newExampleHarness(t)
	scenarios := []struct {
		scope string
		env   map[string]string
		body  string
	}{
		{
			scope: "worktree-only",
			env:   map[string]string{},
			body:  "removed the feature-seed worktree; branch kept",
		},
		{
			scope: "worktree-and-branch",
			env:   map[string]string{},
			body:  "removed the feature-seed worktree and its branch",
		},
		{
			scope: "worktree-and-branch",
			env:   map[string]string{"HWF_E2E_BRANCH_UNMERGED": "1"},
			body:  "removed the feature-seed worktree; branch kept, it is not merged (git branch -D feature-seed)",
		},
	}
	for _, scenario := range scenarios {
		calls := runOK(t, h, "worktree", map[string]string{
			"mode": "delete", "worktree": "feature-seed", "scope": scenario.scope,
		}, scenario.env)
		if findCall(calls, "worktree.create") != nil {
			t.Fatal("unexpected worktree.create")
		}
		shown := notifications(calls)
		if len(shown) != 1 {
			t.Fatalf("scope=%s notifications = %d", scenario.scope, len(shown))
		}
		if fmtString(shown[0]["title"]) != "Worktree deleted" {
			t.Fatalf("title = %v", shown[0]["title"])
		}
		if body, ok := shown[0]["body"].(string); !ok || body != scenario.body {
			t.Fatalf("body = %q want %q", shown[0]["body"], scenario.body)
		}
	}
}

func filterCalls(calls []rpcCall, method string) []rpcCall {
	out := make([]rpcCall, 0)
	for _, call := range calls {
		if call.Method == method {
			out = append(out, call)
		}
	}
	return out
}

func filterCallsWhere(calls []rpcCall, method string, pred func(map[string]any) bool) []rpcCall {
	out := make([]rpcCall, 0)
	for _, call := range calls {
		if call.Method == method && pred(call.Params) {
			out = append(out, call)
		}
	}
	return out
}

func findCall(calls []rpcCall, method string) *rpcCall {
	for i := range calls {
		if calls[i].Method == method {
			return &calls[i]
		}
	}
	return nil
}

func fmtString(v any) string {
	if v == nil {
		return ""
	}
	return strings.TrimSpace(strings.ReplaceAll(fmt.Sprint(v), "\n", ""))
}

func sanitizeAgentName(name string) string {
	var b strings.Builder
	for _, r := range name {
		if (r >= 'A' && r <= 'Z') || (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') || r == '-' {
			b.WriteRune(r)
		} else {
			b.WriteByte('-')
		}
	}
	return b.String()
}
